keps/sig-architecture/2527-clarify-status-observations-vs-rbac/README.md

```diff
@@ -0,0 +1,281 @@
+down APIs that run afoul of the rebuild-from-observation test, and forced the
+implementations to be more complicated.  I no longer find this to be useful to
+our APIs, and in fact I find it to be a detriment.
+
+I suspect that several APIs would have come out differently if not for this.
+
+### Goals
+
+* Clarify or remove the from-observation rule to better match reality.
+* Provide guidance for APIs on how to save controller results.
+
+### Non-Goals
+
+* Retroactively apply this pattern to change existing GA APIs.
+* Necessarily apply this pattern to change pre-GA APIs (though they may choose
+  to follow it).
+* Provide arbitrary storage for controller state.
+* Provide distinct access control for subsets of spec or status.
+
+## Proposal
+
+This KEP proposes to:
+
+1) Get rid of the [if it got lost](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status)
+litmus test for status fields
+2) Acknowledge that spec/status are a useful and meaningful split for access control
+3) Document one or more patterns for APIs that need a trusted acknowledgement
+   of the spec
+
+Depending on feedback, there may be other approaches to solving the same
+problems.
+
+### Risks and Mitigations
+
+The current guidance exists for a reason.  It does encourage a sort of
+ideological cleanliness.  However, it is not universally adhered to and it
+neglects the reality of these request/acknowledge APIs.  On the whole, this KEP
+posits that the current guidance is a net negative.
+
+## Design Details
+
+### Part 1: Loosen and clarify `status`
+
+Remove the idea that status fields _must_ be from observation.  Allow controllers
+to write values to status that represent allocations or acknowledged requests.
+Document that status fields are best when they represent observed state, but do
+not _require_ it.
+
+This has [already happened](https://github.com/kubernetes/enhancements/pull/2308/files#r567809465)
+at least once.
+
+#### Examples
+
+Some of these are variants of existing APIs and some are hypothetical.
+
+1) User creates a Service and sets `Service.spec.type = LoadBalancer`.  The
+cloud controller sets `Service.status.loadBalancer.ingress.ip` in response.  If
+the user sets `Service.spec.loadBalancerIP` to a specific value, the cloud
+controller will either successfully use that IP and and set it as before (ACK),
+or it will not set any value in `Service.status.loadBalancer.ingress` (NAK).
+
+2) Given a Pod, user patches `Pod.spec.containers[0].resources.requests[cpu]`
+to a new value.  Kubelet sees this as a request and, if viable, sets
+`Pod.status.containerStatuses[0].resources.requests[cpu]` to the same value.
+
+3) User creates a `Bucket`, setting `Bucket.spec.instance = artifacts-prod`.
+The bucket controller verifies that the namespace is allowed to use the
+artifacts-prod bucket and, if so, sets `Bucket.status.instance` to the same
+value.
+
```

Comment:
If we're going to recognize or even recommend using status as place for controllers to store state, it would be good to describe patterns / ordering of API status writes w.r.t. taking action (whether on an external system or against a different Kubernetes API object) to ensure a controller is re-entrant and never puts itself in an ambiguous state on restart. 

I've seen controllers get themselves in trouble / become inconsistent or non-reentrant when they hit conflicts or other write errors assuming all operations succeed or fail together doing things like this:
* "act on something else, then write completion to status"
* "write to status in advance, then act on something else"
* "write to status in advance, then act on something else, then write completion to status"

That's not necessarily new to this relaxed definition of status, but "status is from observation" tends to be more one-way, and I expect relaxing that will lead to more of these "back-and-forth" or "write to both" attempts.